pythoncode/FortiusAntBody.py

```diff
@@ -739,102 +735,63 @@ def Tacx2Dongle(self):
-            #       required power may change with speed, for same grade
-            #------------------------------------------------------------------
-            if TargetMode == mode_Grade:
-                r = usbTrainer.Grade2Resistance(TargetGrade, UserAndBikeWeight, SpeedKmh, Cadence)
-                r *= PowercurveFactor
-                TargetPower = usbTrainer.Resistance2Power(r, SpeedKmh)
+                if   TacxTrainer.Buttons == usbTrainer.EnterButton:     pass
+                elif TacxTrainer.Buttons == usbTrainer.DownButton:      TacxTrainer.SetPowercurveFactorDown()
+                elif TacxTrainer.Buttons == usbTrainer.UpButton:        TacxTrainer.SetPowercurveFactorUp()
+                elif TacxTrainer.Buttons == usbTrainer.CancelButton:    self.RunningSwitch = False
+                else:                                       pass
 
             #-------------------------------------------------------------------
-            # Prepare
+            # Do ANT work every 1/4 second
             #-------------------------------------------------------------------
             messages = []       # messages to be sent to ANT
             data = []           # responses received from ANT
-
-            #-------------------------------------------------------------------
-            # Send data to trainer (either power OR grade)
-            #-------------------------------------------------------------------
-            if False and clv.SimulateTrainer:   # I DO NOT REMEMBER NOW
-                pass                            # WHY THIS IS NOT DONE LIKE THIS
-
-            elif clv.Tacx_iVortex and VTX_VortexID:
-                if QuarterSecond:
-                    # Resistance = usbTrainer.Power2Resistance(TargetPower, SpeedKmh, Cadence)
-                    info = ant.msgPage16_TacxVortexSetPower (ant.channel_VTX_s, VTX_VortexID, TargetPower) # Resistance?
-                    messages.append ( ant.ComposeMessage (ant.msgID_BroadcastData, info) )
-
-            elif devTrainer:
-                usbTrainer.SendToTrainer(devTrainer, usbTrainer.modeResistance, \
-                    TargetMode, TargetPower, TargetGrade, PowercurveFactor, \
-                    UserAndBikeWeight, \
-                    PedalEcho, SpeedKmh, Cadence, Calibrate, clv.SimulateTrainer)
-
             if QuarterSecond:
                 LastANTtime = time.time()
-                #-------------------------------------------------------------------
+                #---------------------------------------------------------------
+                # Send power to iVortex
+                # Yes, we could have put this into the clsTacxTrainer
+                #      now all the ANT traffic is here.
+                #---------------------------------------------------------------
+                if clv.Tacx_iVortex and VTX_VortexID:
+                    info = ant.msgPage16_TacxVortexSetPower (ant.channel_VTX_s, VTX_VortexID, TacxTrainer.TargetPower)

```

Comment:
Either here, or inside msgPage16_TacxVortexSetPower... don't let TargetPower go negative - clip to 0 as unsigned short is used in the packed format.

Reply:
change into short

Reply:
Is i-Vortex powered? Can it go -ve?

Reply:
I will publish Vortex headunit shortly. Then we're fully on

Reply:
Two approaches... change unsigned short to short, or leave unsigned short but clip -ve values to 0.

Which is correct? Probably answered by whether i-Vortex can simulate -ve grades - is it a motor brake or a magnetic resistance unit? If it's a motor then signed short, if it's just a brake, then clip at 0... what's correct?

Reply:
My assumption is that iVortex is newer than Fortius and hence will support -Power.
But it remains an assumption.

Reply:
https://tacx.com/product/i-vortex/
Max. incline 7%
Simulation of descent No

Reply:
Thanks for checking Power >= 0

Reply:
Did you put a fix in for this yet? Was looking for it but didn't see it.

Reply:
def msgPage16_TacxVortexSetPower (Channel, VortexID, Power):
    DataPageNumber      = 16

    fChannel            = sc.unsigned_char  # First byte of the ANT+ message content
    fDataPageNumber     = sc.unsigned_char  # First byte of the ANT+ datapage (payload)
    fVortexID           = sc.unsigned_short
    fCommand            = sc.unsigned_char  # 0xAA power request
    fSubcommand         = sc.unsigned_char
    fNoCalibrationData  = sc.unsigned_char
    fPower              = sc.unsigned_short # https://tacx.com/nl/product/i-vortex/
    **Power = min(0, Power)                   # --> No simulation descent ==> power > 0**

    format = sc.big_endian +    fChannel + fDataPageNumber +   fVortexID + fCommand + fSubcommand + fNoCalibrationData + fPower
    info   = struct.pack(format, Channel,   DataPageNumber, int(VortexID),  0xAA,      0,            0,               int(Power))

    return info

Reply:
Indeed, I missed it. Thank you.

Unless I've forgotten to follow up on anything I had in mind, thumbs up from me for this being candidate release. :D

Reply:
Hold on.... min(0,-1) is -1. Do you mean max()?

Reply:
sorry, sure I do

Reply:
uploaded